proposals/0057-events.md

```diff
@@ -0,0 +1,261 @@
+
+A lot of developers would prefer a reliable and universally compatible solution
+so recently Anchor has added a new feature called cpi_events. This is a
+convention on how to call your own program via CPI and encode a binary string
+as instruction data. It can be indexed via the RPC Transaction response, as
+innerInstruction data is always persisted. It is more reliable than sol_log_
+but it effectively reduces runtime limits like maximum CPI call depth for the
+actual application. It also causeses unnecessary overhead in replay as a new
+execution context neeeds to be created, for something as trivial as a log
+message. For the average program (1-2mb) that would be around 5000-9000 CU per
+1kb log message.
+
+In comparison to ethereum and cosmos, solana's preferred access for transaction
+meta-data is via address filtering, e.g. `getSignaturesForAddress` call.
+
+1. Topics: the ethereum yellow paper instead specifies that every event can be
+attributed to up to 4 topics. Each topic is a 32 byte subscription key, users
+can use to filter these events via RPC.
+
+This seems restrictive compared to the expressivenes that is guaranteed through
+the context of each instruction, program id and account metas, this data is
+anyways available at the RPC node.
+
+2. Schema: the cosmos sdk requires every event to have a protobuf schema, this
+makes it very easy for third parties to come in and build generalized indexers
+for the cosmos ecosystem. It also allows to specify complex filter queries that
+can evaluated on the RPC node to stream or retrieve events.
+
+Parsing the event payload is not a task for the rpc layer, and should instead
+be handled by 3rd party providers. As schemas on solana are flexible and can
+change with program updates, this would require access to historic schema
+definitions / account state.
+
+## New Terminology
+
+Discriminator - a unique n byte sequence often a 8 byte hash of a method
+signature or data type, that should be used to tag the type of a data structure
+and encoded always in the very first bytes of an event, account or instruction
+for the sake of efficient parsing.
+
+POA - Program Owned Account
+
+EOA - Externally Owned Account
+
+## Detailed Design
+
+### Runtime 
+
+In order to log an event, the SVM runtime must implement a new syscall
+`sol_emit_`. The runtime must charge 100 + ceil(48 * len / 1024) CU per
+invocation.
+
+```
+#include <sol/emit.h>
+void sol_emit_(/* r1 */ uint64_t len,
+               /* r2 */ const char *message);
+```
+
+The validator should preserve these events similar to transaction history
+and deliver them to it's operators via RPC & Geyser interfaces. Although
+those are technically not part of the SIMD specification process, it's
+important to publish a recommended spec for 3rd party providers so that Solana
+Dapps don't run into vendor lock-in issues.
+
+### RPC-Client
+
+Clients should be able subscribe to events emitted with various filters:
+
+```
+rpc.onEvents(
```

Comment:
There should be some way for the client to know the slot height that their subscription starts at. This'll allow them to backfill any missed events prior to their subscription using some other method.

Reply:
it should work well to wait for the first event an then backfill (the event includes the slot)
in the mean-time the user can manually query getSlot after opening the subscription and backfill from there

Reply:
waiting for the first event might not work well for events that don't happen frequently though right?

Reply:
yeah but slot updates itself are frequent enough imo